File: my_daa_examples/bfs.py
###############################################################################
#
# \file    bfs.py
# \author  Sudnya Padalikar <mailsudnya@gmail.com>
# \date    Saturday Aug 27, 2016
# \brief   breadth first search
#
###############################################################################

import argparse
import logging
from collections import deque

logger = logging.getLogger("BFS")

def runBfs(graph, startingNode):
    frontier = deque()
    frontier.append(startingNode)

    visited = set()

    while len(frontier) != 0:
        logger.info("Frontier contains: " + str(frontier))
        logger.info("visited contains: " + str(visited))
        node = frontier.popleft()
        logger.info ("BFS traverse: " + str(node))
        
        visited.add(node)
        adjNodes = graph.get(node)
        if adjNodes == None:
            continue

        for entry in adjNodes:
            if entry not in visited:
                logger.info("Adding new (unvisited) connected node: " + str(entry) + " to frontier list")
                frontier.append(entry)
            else:
                logger.info("Already saw: " + str(entry) + " do nothing")


def createGraph():
    adjList = {}
    adjList[0] = set([1,2])
    adjList[1] = set([2,3])
    adjList[2] = set([5])
    adjList[3] = set([4])
    adjList[4] = set([1])
    adjList[5] = None

    return adjList

def main():
    parser = argparse.ArgumentParser(description="Implementing BFS ")
    parser.add_argument("-v", "--verbose", default = False, action = "store_true")

    parsedArguments = parser.parse_args()
    arguments = vars(parsedArguments)

    isVerbose   = arguments['verbose']

    if isVerbose:
        logging.basicConfig(level=logging.DEBUG)
    else:
        logging.basicConfig(level=logging.INFO)
    graph = createGraph()
    runBfs(graph, 0)

if __name__ == '__main__':
    main()

File: my_daa_examples/bfs.cpp
/*
 *
 * \file    bfs.cpp
 * \author  Sudnya Diamos <mailsudnya@gmail.com>
 * \date    Friday Dec 9, 2016
 * \brief   breadth first search
 *
**/

#include <iostream>
#include <vector>
#include <set>
#include <deque>

template <typename T>
void printSet(const T v)
{
    for (auto& i : v)
        std::cout << i << " , ";
    std::cout << "\n";
}

template <typename T>
void runBfs(T g) 
{
    std::set<int> visited;
    std::deque<int> frontier;
    
    frontier.push_back(0);
    visited.insert(0);

    while(!frontier.empty() && visited.size() != g.size() )
    {
        //get adjacent
        auto current = frontier.front();
        std::cout << "visiting current node: " << current << std::endl;
        frontier.pop_front();
        
        visited.insert(current);

        auto adjacent = g.at(current);

        for (auto i = adjacent.begin(); i != adjacent.end(); ++i)
        {
            bool isVisited = visited.find(*i) != visited.end();
            if (!isVisited)
            {
                std::cout << "adding " << *i << " to frontier\n";
                frontier.push_back(*i);
            }
        }
        std::cout << "frontier: ";
        printSet(frontier);
        std::cout << "visited: ";
        printSet(visited);
    }
}

template <typename T>
void printInput(T g)
{
    int counter = 0;
    for (auto i = g.begin(); i != g.end(); ++i, ++counter)
    {
        std::cout << "\nNode: " << counter << " : ";
        for (auto j = i->begin(); j != i->end(); ++j)
        {
            std::cout << *j << " \t ";
        }
    }

}
/*
 *      0 -> 1 ->   3
 *     \|  .  /.\  \|/
 * 5 <- 2 \./     .4
 *
 */
int main()
{
    std::cout << "BFS\n";
    //create a graph
    //0-> 1, 2
    //1-> 2, 3
    //2-> 5
    //3-> 4
    //4-> 1
    std::vector<int>zero  = {1,2};
    std::vector<int>one   = {2,3};
    std::vector<int>two   = {5};
    std::vector<int>three = {4};
    std::vector<int>four  = {1};
    std::vector<int>five  = {};

    std::vector<std::vector<int>> g = {zero, one, two, three, four, five};

    //printInput(g);

    runBfs(g);
}
